crates/red_knot_python_semantic/src/semantic_index/usedef.rs

```diff
@@ -0,0 +1,127 @@
+use crate::ast_node_ref::AstNodeRef;
+use crate::db::Db;
+use crate::semantic_index::ast_ids::ScopedUseId;
+use crate::semantic_index::definition::Definition;
+use crate::semantic_index::symbol::{FileScopeId, ScopeId, ScopedSymbolId};
+use ruff_db::files::File;
+use ruff_index::IndexVec;
+use ruff_python_ast as ast;
+use salsa;
+
+/// An independently type-inferable expression.
+///
+/// Includes constraint expressions (e.g. if tests) and the RHS of an unpacking assignment.
+#[salsa::tracked]
+pub(crate) struct Expression<'db> {
+    /// The file in which the expression occurs.
+    #[id]
+    pub(crate) file: File,
+
+    /// The scope in which the expression occurs.
+    #[id]
+    pub(crate) file_scope: FileScopeId,
+
+    /// The statement node containing the expression.
+    /// We hold the statement node to clarify which specific contexts these
+    /// independently-inferable expressions can occur in, and limit the variants of this enum.
+    #[no_eq]
+    #[return_ref]
+    pub(crate) ast: ExpressionKind,
+}
+
+impl<'db> Expression<'db> {
+    /// Return the actual expression.
+    #[allow(unused)]
+    pub(crate) fn node(self, db: &'db dyn Db) -> &ast::Expr {
+        match self.ast(db) {
+            ExpressionKind::AssignedValue(node_ref) => &node_ref.node().value,
+            ExpressionKind::IfTest(node_ref) => &node_ref.node().test,
+        }
+    }
+
+    pub(crate) fn scope(self, db: &'db dyn Db) -> ScopeId<'db> {
+        self.file_scope(db).to_scope_id(db, self.file(db))
+    }
+}
+
+#[derive(Clone, Debug)]
+#[allow(dead_code)]
+pub(crate) enum ExpressionKind {
+    AssignedValue(AstNodeRef<ast::StmtAssign>),
+    IfTest(AstNodeRef<ast::StmtIf>),
+    // TODO if-expressions, loops, etc
+}
+
```

Comment:
I'm a bit surprised to find the `Expression` ingredient in this module. What's the motivation for placing it here rather than e.g. in its own module?

Reply:
Laziness :) I'll move it to its own module.